cardinal/ecs/systems/manager.go

```diff
@@ -0,0 +1,175 @@
+	"reflect"
+	"runtime"
+	"time"
+)
+
+type System func(ctx engine.Context) error
+
+type Manager struct {
+	// registeredSystems is a list of all the registered system names in the order that they were registered.
+	// This is represented as a list as maps in Go are unordered.
+	registeredSystems []string
+
+	// systemFn is a map of system names to system functions.
+	systemFn map[string]System
+
+	// currentSystem is the name of the system that is currently running.
+	currentSystem *string
+
+	initSystem      System
+	isInitSystemRan bool
+}
+
+// New creates a new system manager.
+func New() *Manager {
+	return &Manager{
+		registeredSystems: make([]string, 0),
+		systemFn:          make(map[string]System),
+		currentSystem:     nil,
+	}
+}
+
+// RegisterSystems registers multiple systems with the system manager.
+// There can only be one system with a given name, which is derived from the function name.
+// If there is a duplicate system name, an error will be returned and none of the systems will be registered.
+func (m *Manager) RegisterSystems(systems ...System) error {
+	systemNames := make([]string, 0, len(systems))
+
+	// Iterate through all the systems and check if they are already registered.
+	// This is done before registering any of the systems to ensure that all systems are registered or none of them are.
+	for _, system := range systems {
+		// Obtain the name of the system function using reflection.
+		systemName := filepath.Base(runtime.FuncForPC(reflect.ValueOf(system).Pointer()).Name())
+		fmt.Println(systemName)
+
+		// Checks if the system is already previously registered.
+		// This will terminate the registration of all systems if any of them are already registered.
+		if err := m.checkDuplicateSystemName(systemName); err != nil {
+			return eris.Wrap(err, "failed to register system")
+		}
+
+		// If the system is not already registered, add it to the list of system names.
+		systemNames = append(systemNames, systemName)
+	}
+
+	// Iterate through all the systems and register them one by one.
+	for i, systemName := range systemNames {
+		m.registeredSystems = append(m.registeredSystems, systemName)
+		m.systemFn[systemName] = systems[i]
+	}
+
+	return nil
+}
+
+// RegisterInitSystem registers an init system with the system manager.
+// The init system can only be run once.
+func (m *Manager) RegisterInitSystem(system System) {
+	m.initSystem = system
+}
+
+// RunSystems runs all the registered system in the order that they were registered.
+func (m *Manager) RunSystems(eCtx engine.Context) error {
+	allSystemStartTime := time.Now()
+
+	for _, systemName := range m.registeredSystems {
+		m.currentSystem = &systemName
+
+		// Inject the system name into the logger
+		eCtx.SetLogger(eCtx.Logger().With().Str("system", systemName).Logger())
+
+		// Executes the system function that the user registered
+		systemStartTime := time.Now()
+		err := m.systemFn[systemName](eCtx)
+		if err != nil {
+			m.currentSystem = nil
```

Comment:
unfortunately this will break the expected behavior. The current system is currently used to indicate what system is last running when a panic (from within the system) occurs. If we add a defer, then it will always be nil if there's a panic.